akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala

```diff
@@ -242,5 +242,96 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
         assert(calculatedValue === expectedResult)
       }
     }
+
+    "stop restarting the child after reaching maxNrOfRetries limit (Backoff.onStop)" in {
+      val supervisor = create(onStopOptions(maxNrOfRetries = 2))
+      def waitForChild(n: Int = 0): Option[ActorRef] = {
+        supervisor ! BackoffSupervisor.GetCurrentChild
+        val c = expectMsgType[BackoffSupervisor.CurrentChild].ref
+        if (c.isDefined)
+          c
+        else if (n < 20) {
+          Thread.sleep(50)
+          waitForChild(n + 1)
+        } else None
+      }
+
+      watch(supervisor)
+
+      supervisor ! BackoffSupervisor.GetRestartCount
+      expectMsg(BackoffSupervisor.RestartCount(0))
+
+      supervisor ! BackoffSupervisor.GetCurrentChild
+      val c1 = expectMsgType[BackoffSupervisor.CurrentChild].ref.get
+      watch(c1)
+      c1 ! PoisonPill
+      expectTerminated(c1)
+
+      supervisor ! BackoffSupervisor.GetRestartCount
+      expectMsg(BackoffSupervisor.RestartCount(1))
+
+      val c2 = waitForChild().get
+      awaitAssert(c2 should !==(c1))
+      watch(c2)
+      c2 ! PoisonPill
+      expectTerminated(c2)
+
+      supervisor ! BackoffSupervisor.GetRestartCount
+      expectMsg(BackoffSupervisor.RestartCount(2))
+
+      val c3 = waitForChild().get
+      awaitAssert(c3 should !==(c2))
+      watch(c3)
+      c3 ! PoisonPill
+      expectTerminated(c3)
+      expectTerminated(supervisor)
+    }
+
+    "stop restarting the child after reaching maxNrOfRetries limit (Backoff.onFailure)" in {
+      filterException[TestException] {
+        val supervisor = create(onFailureOptions(maxNrOfRetries = 2))
+
+        def waitForChild(n: Int = 0): Option[ActorRef] = {
+          supervisor ! BackoffSupervisor.GetCurrentChild
+          val c = expectMsgType[BackoffSupervisor.CurrentChild].ref
+          if (c.isDefined)
+            c
+          else if (n < 20) {
+            Thread.sleep(50)
+            waitForChild(n + 1)
+          } else None
+        }
+
+        watch(supervisor)
+
+        supervisor ! BackoffSupervisor.GetRestartCount
+        expectMsg(BackoffSupervisor.RestartCount(0))
+
+        supervisor ! BackoffSupervisor.GetCurrentChild
+        val c1 = expectMsgType[BackoffSupervisor.CurrentChild].ref.get
+        watch(c1)
+        c1 ! "boom"
+        expectTerminated(c1)
+
+        supervisor ! BackoffSupervisor.GetRestartCount
+        expectMsg(BackoffSupervisor.RestartCount(1))
+
+        val c2 = waitForChild().get
```

Comment:
Perhaps you could use an `eventually` block instead of the custom `waitForChild` here?

Reply:
Done